metrics/src/receiver.rs

```diff
@@ -1,220 +1,73 @@
 use crate::{
-    configuration::Configuration,
-    control::{ControlFrame, Controller},
-    data::{Counter, Gauge, Histogram, Sample, ScopedKey, Snapshot, StringScopedKey},
-    scopes::Scopes,
+    builder::{Builder, BuilderError},
+    common::MetricScope,
+    config::MetricConfiguration,
+    control::Controller,
+    registry::{MetricRegistry, ScopeRegistry},
     sink::Sink,
 };
-use crossbeam_channel::{self, bounded, tick, Select, TryRecvError};
-use quanta::Clock;
-use std::{
-    sync::Arc,
-    time::{Duration, Instant},
-};
-
-/// Wrapper for all messages that flow over the data channel between sink/receiver.
-pub(crate) enum MessageFrame {
-    /// A normal data message holding a metric sample.
-    Data(Sample),
-}
-
-/// Metrics receiver which aggregates and processes samples.
+use quanta::{Builder as UpkeepBuilder, Clock, Handle as UpkeepHandle};
+use std::sync::Arc;
+use std::time::Duration;
+
+/// Central store for metrics.
+///
+/// `Receiver` is the nucleus for all metrics operations.  While no operations are performed by it
+/// directly, it holds the registeries and references to resources and so it must live as long as
+/// any [`Sink`] or `[`Controller`] does.
 pub struct Receiver {
-    config: Configuration,
-
-    // Sample aggregation machinery.
-    msg_tx: crossbeam_channel::Sender<MessageFrame>,
-    msg_rx: Option<crossbeam_channel::Receiver<MessageFrame>>,
-    control_tx: crossbeam_channel::Sender<ControlFrame>,
-    control_rx: Option<crossbeam_channel::Receiver<ControlFrame>>,
-
-    // Metric machinery.
-    counter: Counter,
-    gauge: Gauge,
-    thistogram: Histogram,
-    vhistogram: Histogram,
-
+    metric_registry: Arc<MetricRegistry>,
+    scope_registry: Arc<ScopeRegistry>,
     clock: Clock,
-    scopes: Arc<Scopes>,
+    _upkeep_handle: UpkeepHandle,
```

Comment:
is this still needed?

Reply:
The `UpkeepHandle` is essentially a newtype around `std::thread::JoinHandle`, and so we need to hold it to keep the background Quanta time thread running.  I only prefixed it to get the compiler to shut up about it.

Reply:
👍